### PR TITLE
System indices ignore all user templates

### PR DESCRIPTION
**Pull request by williamrandolph**
System indices should ignore all user templates, except when explicitly stated (allowTemplates flag on SystemIndexDescriptor). This was the intended behavior for 8.0.

If merged, this will fix the longstanding issue #42508, as well as the stepping-stone issue #74271.

- [x] Verify that Kibana is not using templates for its settings and mappings

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -632,6 +639,52 @@ private ClusterState applyCreateIndexRequestWithV2Template(
         );
     }
 
+    private ClusterState applyCreateIndexRequestForSystemIndex(
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:642]: At first, I used applyCreateIndexRequestWithV1Templates and a list of empty templates, but that seemed very unsatisfying.

**Comment by grcevski** [comment]
Kibana does use templates for at least one system index `.kibana_security_session_1`

https://github.com/elastic/kibana/blob/3730dd0779ed8efd74aee88f57422781ec1ac122/x-pack/plugins/security/server/session_management/session_index.ts#L61-L92

Alternatively we can ignore user templates for all MANAGED system indices.

**Comment by grcevski** [comment]
Here's the related Kibana issue to migrate the security_session indices to supply the index options at creation time: https://github.com/elastic/kibana/issues/134897

**Comment by grcevski** [comment]
To mitigate the fact that Kibana is using templates we introduced the allowsTemplates flag on SystemIndexDescriptor (https://github.com/elastic/elasticsearch/pull/87933). We now use this flag in MetadataCreateIndexService to only allow templates to impact index creation on index descriptors that have this flag. 

Tests have been adjusted accordingly to reflect this behaviour. We ensure that templates have impact only on index descriptors that have the flag explicitly on.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java [hunk 2]
```diff
@@ -167,40 +168,60 @@ public void testSystemIndicesAutoCreateRejectedWhenNotHidden() {
...
+        assertAliasesHidden(
+            nonPrimaryIndex,
+            Set.of(
+                TestSystemIndexDescriptorAllowsTemplates.INDEX_NAME,
+                TestSystemIndexDescriptorAllowsTemplates.INDEX_NAME + "-legacy-alias"
+            ),
+            2
+        );
+
+        assertAcked(client().admin().indices().prepareDeleteTemplate("*").get());
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java:215]: This line isn't necessary, it's done in the `@After` (and this should be handled by the test infrastructure anyway, I thought).

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java [hunk 3]
```diff
@@ -167,40 +168,60 @@ public void testSystemIndicesAutoCreateRejectedWhenNotHidden() {
...
+
+    /**
+     * Check that a legacy template does not create an alias for a system index
+     */
+    public void testAutoCreateSystemAliasViaV1Template() throws Exception {
+        var nonPrimaryIndex = autoCreateSystemAliasViaV1Template(INDEX_NAME);
+
+        assertAliasesHidden(nonPrimaryIndex, Set.of(INDEX_NAME), 1);
 
         assertAcked(client().admin().indices().prepareDeleteTemplate("*").get());
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java:197]: Ah, it looks like this line was copied - can you remove it from here as well? It's equally unnecessary.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java [hunk 4]
```diff
@@ -157,37 +158,60 @@ public void testCreateSystemIndexViaConcreteName() {
...
+            Set.of(
+                TestSystemIndexDescriptorAllowsTemplates.INDEX_NAME,
+                TestSystemIndexDescriptorAllowsTemplates.INDEX_NAME + "-legacy-alias"
+            ),
+            TestSystemIndexDescriptorAllowsTemplates.INDEX_NAME,
+            TestSystemIndexDescriptorAllowsTemplates.PRIMARY_INDEX_NAME,
+            2
+        );
+
+        assertAcked(client().admin().indices().prepareDeleteTemplate("*").get());
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:205]: Same here, this is done in the `@After`.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java [hunk 5]
```diff
@@ -157,37 +158,60 @@ public void testCreateSystemIndexViaConcreteName() {
...
+    /**
+     * Check that a legacy template does not create an alias for a system index
+     */
+    public void testCreateSystemAliasViaV1Template() throws Exception {
+        createSystemAliasViaV1Template(INDEX_NAME, PRIMARY_INDEX_NAME);
 
-        assertHasAliases(Set.of(".test-index", ".test-index-legacy-alias"));
+        assertHasAliases(Set.of(INDEX_NAME), INDEX_NAME, PRIMARY_INDEX_NAME, 1);
 
         assertAcked(client().admin().indices().prepareDeleteTemplate("*").get());
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:183]: Another one.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/TestSystemIndexDescriptorAllowsTemplates.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+package org.elasticsearch.indices;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A special kind of {@link SystemIndexDescriptor} that can toggle what kind of mappings it
+ * expects. A real descriptor is immutable.
+ */
+public class TestSystemIndexDescriptorAllowsTemplates extends TestSystemIndexDescriptor {
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/TestSystemIndexDescriptorAllowsTemplates.java:17]: Can you rename this class? If the javadoc is right, the special thing isn't that it allows templates, it's that the mapping is switchable.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java [hunk 7]
```diff
@@ -214,13 +235,45 @@ public void testAutoCreateSystemAliasViaComposableTemplate() throws Exception {
...
+
+    /**
+     * Check that a composable template does not create an alias for a system index
+     */
+    public void testAutoCreateSystemAliasViaComposableTemplate() throws Exception {
+        String nonPrimaryIndex = autoCreateSystemAliasViaComposableTemplate(INDEX_NAME);
+
+        assertAliasesHidden(nonPrimaryIndex, Set.of(INDEX_NAME), 1);
+
+        assertAcked(
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java:255]: I'm pretty sure this is unnecessary as well despite not being explicitly in the `@After`, I ran it locally without this x10 and every thing, passed, and `:server:check` passed as well.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java [hunk 8]
```diff
@@ -201,10 +225,44 @@ public void testCreateSystemAliasViaComposableTemplate() throws Exception {
...
-        assertHasAliases(Set.of(".test-index", ".test-index-composable-alias"));
+    /**
+     * Check that a composable template does not create an alias for a system index
+     */
+    public void testCreateSystemAliasViaComposableTemplate() throws Exception {
+        createIndexWithComposableTemplates(INDEX_NAME, PRIMARY_INDEX_NAME);
+
+        assertHasAliases(Set.of(INDEX_NAME), INDEX_NAME, PRIMARY_INDEX_NAME, 1);
+
+        assertAcked(
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:240]: I haven't tried it locally but I bet we don't need this one either.

**Review by AthenaEryma — COMMENTED** [review]
Left a bunch of comments but they're all really minor, this looks really good!

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java (hunk 7 above)
Comment (grcevski) [reply server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java:255]: I'm inclined to leave it, it does clean up the template explicitly and at the same time it tests we can clean up composable templates.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java (hunk 8 above)
Comment (grcevski) [reply server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:240]: I think just like the other one, it doesn't hurt.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/TestSystemIndexDescriptorAllowsTemplates.java (hunk 6 above)
Comment (grcevski) [reply server/src/internalClusterTest/java/org/elasticsearch/indices/TestSystemIndexDescriptorAllowsTemplates.java:17]: The javadoc is wrong, I'll fix it. The purpose of the extra descriptor is to be doing exactly the same as the other test descriptor, but also allow templates to apply to it like the kibana-sessions templates.

**Comment by grcevski** [comment]
@elasticmachine update branch

**Comment by grcevski** [comment]
Hey @gwbrown, I think I addressed all of your feedback. Do you mind taking another look?

**Comment by legrego** [comment]
> Here's the related Kibana issue to migrate the security_session indices to supply the index options at creation time: https://github.com/elastic/kibana/issues/134897

The fix for this has been merged to Kibana's 'main' branch.

**Review by AthenaEryma — APPROVED** [review]
LGTM, thanks!

**Pull request merged**